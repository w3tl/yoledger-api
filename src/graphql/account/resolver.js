import { GraphQLError } from 'graphql';

export const Account = {
  id: ({ _id }) => _id.toString(),
};

export const Query = {
  async accounts(root, { type }, { models: { accountModel } }) {
    return accountModel.find({ type }).toArray();
  },
  async account(root, { id }, { models: { accountModel } }) {
    return accountModel.findById(id);
  },
};

export const Mutation = {
  async addAccount(root, { input }, { models: { accountModel } }) {
    const { name, type, balance } = input;
    const existsAccount = await accountModel.findOne({ name, type });
    if (existsAccount && existsAccount.type === type) {
      throw new GraphQLError(`${name} already exists`);
    }
    const { insertedId } = await accountModel.create({
      name,
      balance,
      type,
    });

    return {
      account: await accountModel.findOne({ _id: insertedId }),
    };
  },
};
