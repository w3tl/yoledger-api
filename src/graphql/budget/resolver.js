export const Budget = {
  id: ({ _id }) => _id.toString(),
  account({ account }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ ...account, userId });
  },
};

export const Query = {
  async budgets(root, { dateStart, dateEnd }, { models: { budgetModel } }) {
    const accounts = await budgetModel.aggregate([
      {
        $match: { $and: [{ date: { $gte: dateStart } }, { date: { $lte: dateEnd } }] },
      },
      {
        $group: {
          _id: { account: '$account' },
        },
      },
      {
        $project: {
          account: '$_id.account',
        },
      },
      {
        $replaceRoot: { newRoot: '$account' },
      },
    ]).toArray();
    return { accounts };
  },
  async budget(root, { account, dateStart, dateEnd }, { models: { budgetModel } }) {
    const budget = await budgetModel.find({
      'account.name': account,
      $and: [{ date: { $gte: dateStart } }, { date: { $lte: dateEnd } }],
    }).toArray();
    return budget;
  },
};

export const Mutation = {
  async upsertBudget(root, { input }, { models: { budgetModel } }) {
    const { account, date, amount } = input;
    const ok = await budgetModel.upsertBudget({ account, date, amount });
    const budget = await budgetModel.findOne({ 'account.name': account, date });
    return {
      success: !!ok,
      budget,
    };
  },
};
