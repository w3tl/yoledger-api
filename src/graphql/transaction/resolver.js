import { ObjectId } from 'mongodb';
import { GraphQLError } from 'graphql';

export const Transaction = {
  id: ({ _id }) => _id.toString(),
  date: ({ createdAt }) => createdAt,
  source({ source }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ ...source, userId });
  },
  destination({ destination }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ ...destination, userId });
  },
};

export const Mutation = {
  async addTransaction(root, { input }, { models: { transactionModel }, connection }) {
    const {
      source, destination, amount, date: createdAt,
    } = input;
    const transObj = {
      source,
      destination,
      amount,
      createdAt,
    };

    let _id;
    await connection.withSession(async (session) => {
      _id = await transactionModel.post(transObj, session);
    });

    const transaction = await transactionModel.findOne({ _id });

    return {
      transaction,
    };
  },
  async deleteTransaction(root, { id }, { models: { transactionModel }, connection }) {
    try {
      await connection.withSession(session => transactionModel.unpost(ObjectId(id), session));
    } catch (err) {
      return {
        success: false,
      };
    }

    return {
      success: true,
      id,
    };
  },
  async updateTransaction(root, { id, input }, { models: { transactionModel }, connection }) {
    const {
      amount, source, destination, date: createdAt,
    } = input;
    const updates = {
      source,
      destination,
      amount,
      createdAt,
    };
    try {
      let updatedId;
      await connection.withSession(async (session) => {
        updatedId = await transactionModel.update(ObjectId(id), updates, session);
      });
      const transaction = await transactionModel.findOne({ _id: updatedId });
      return { transaction };
    } catch (err) {
      throw new GraphQLError(err.message);
    }
  },
};

export const Query = {
  async transactions(_, /* eslint-disable-next-line object-curly-newline */
    { dateStart, dateEnd, page, itemsPerPage },
    { models: { transactionModel } }) {
    const query = {};
    if (dateEnd) {
      query.$and = [
        { createdAt: { $gte: dateStart } },
        { createdAt: { $lte: dateEnd } },
      ];
    } else {
      query.createdAt = { $gte: dateStart };
    }

    return transactionModel.find(query)
      .skip(page * itemsPerPage)
      .limit(itemsPerPage)
      .toArray();
  },
  transaction(_, { id }, { models: { transactionModel } }) {
    return transactionModel.findById(id);
  },
};
