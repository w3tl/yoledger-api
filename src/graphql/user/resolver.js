import { AuthenticationError, UserInputError } from 'apollo-server-express';
import jwt from 'jsonwebtoken';

const createToken = (userId, secret, expiresIn) => jwt.sign({
  userId,
}, secret, {
  expiresIn,
});

export const User = {
  username: ({ _id }) => _id.toString(),
};

export const Query = {
  user(root, _, { models: { userModel } }) {
    return userModel.findOne();
  },
};

export const Mutation = {
  async signin(root, { login, password }, { models: { userModel }, secret, expiresIn }) {
    let isValid = false;
    try {
      isValid = await userModel.validatePassword(login, password);
    } catch (e) {
      throw new UserInputError('No user found with this login credentials.');
    }

    if (!isValid) {
      throw new AuthenticationError('Invalid password.');
    }

    return {
      token: createToken(login, secret, expiresIn),
    };
  },
  async signout(root, variables, { models: { userModel } }) {
    const user = await userModel.findOne();
    if (!user) {
      throw new UserInputError('No user found with this login');
    }
    return { success: true };
  },
  async changePassword(root, input, {
    models: { userModel }, userId, secret, expiresIn,
  }) {
    const { oldPassword, newPassword } = input;
    let isValid = false;
    try {
      isValid = await userModel.validatePassword(userId, oldPassword);
    } catch (e) {
      throw new UserInputError('No user found with this login credentials.');
    }

    if (!isValid) {
      throw new AuthenticationError('Invalid password.');
    }

    const result = await userModel.changePassword(newPassword);
    if (!result) {
      throw new UserInputError('Password already used');
    }

    return {
      token: createToken(userId, secret, expiresIn),
    };
  },
};
