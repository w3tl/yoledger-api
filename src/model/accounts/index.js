import Model from '../Model';

export const accountTypes = ['ASSET', 'EXPENSE', 'INCOME', 'VENDOR'];

const schema = {
  type: 'object',
  required: ['name', 'type', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    name: {
      type: 'string',
    },
    balance: {
      type: 'number',
    },
    type: {
      enum: accountTypes,
    },
    userId: {
      type: 'string',
    },
  },
};

export default class Account extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.schema = schema;
    this.collName = 'accounts';
  }

  create({
    name, type, balance = 0,
  }) {
    return super.insertOne({
      name,
      type,
      balance,
    });
  }

  addAmount(name, amount, session) {
    return super.updateOne(
      { name, userId: this.userId },
      { $inc: { balance: amount } },
      { session },
    );
  }

  findByName(name, ...props) {
    return super.findOne({ name }, ...props);
  }
}
