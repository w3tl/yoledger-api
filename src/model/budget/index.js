import Model from '../Model';
import Account from '../accounts';
import schema from './schema';

export default class Budget extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.collName = 'budget';
    this.schema = schema;
  }

  async upsertBudget({ account, date, amount }) {
    const accountModel = new Account(this.db, this.userId);
    const accountDoc = await accountModel.findByName(account, {
      projection: { _id: 0, name: 1, type: 1 },
    });
    if (!accountDoc) {
      throw new Error(`Account "${account}" not found`);
    }
    const periodDate = typeof date === 'string'
      ? new Date(date) : date;
    const { upsertedId, modifiedCount } = await super.updateOne({
      date: periodDate, account: accountDoc,
    }, {
      $set: { amount },
    }, {
      upsert: true,
    });
    return upsertedId || modifiedCount;
  }
}
