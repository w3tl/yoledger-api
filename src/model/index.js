import Account from './accounts';
import Transaction from './transactions';
import Budget from './budget';
import User from './user';

export {
  User,
  Account,
  Transaction,
  Budget,
};
