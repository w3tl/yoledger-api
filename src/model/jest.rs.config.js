module.exports = {
  testEnvironment: 'node',
  notify: true,
  transform: {
    '^.+\\.js?$': 'babel-jest',
  },
  globalSetup: './setup.js',
  globalTeardown: './teardown.js',
};
