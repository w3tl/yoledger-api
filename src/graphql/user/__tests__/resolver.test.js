jest.mock('mongodb');
jest.mock('jsonwebtoken');

import { Query, Mutation, User } from '../resolver';
import { User as UserModel } from '../../../model';

const userId = 'admin';

describe('user resolver', () => {
  const { connection } = require('mongodb');
  const models = { userModel: new UserModel(connection.db(), userId) };
  test('Query', async () => {
    const result = await Query.user(null, {}, { models });
    expect(result).toMatchSnapshot();
  });

  test('User', () => {
    expect(User.username({ _id: userId })).toBe(userId);
  });

  describe('Mutation', () => {
    test('signin should return token for correct credentials', async () => {
      const context = { models, secret: 'secret', expiresIn: 10 };
      const result = await Mutation.signin(null, { login: userId, password: 'password' }, context);
      expect(result).toMatchSnapshot();
    });

    describe('changePassword', () => {
      test('should throw an error when password are used', async () => {
        const context = {
          models: { userModel: new UserModel(connection.db(), 'user1') },
          secret: 'secret',
          expiresIn: 10,
          userId: 'user1',
        };
        await expect(Mutation.changePassword(null, { oldPassword: 'password', newPassword: 'passwordUsed' }, context))
          .rejects
          .toThrowError(/Password already used/);
      });

      test('should return token when when password was changed', async () => {
        const context = {
          models, secret: 'secret', expiresIn: 10, userId: 'admin',
        };
        const result = await Mutation.changePassword(null, { oldPassword: 'password', newPassword: 'newPassword' }, context);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
