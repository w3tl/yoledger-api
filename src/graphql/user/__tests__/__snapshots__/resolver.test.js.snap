// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resolver Mutation changePassword should return token when when password was changed 1`] = `
Object {
  "token": "{\\"userId\\":\\"admin\\"};secret;{\\"expiresIn\\":10}",
}
`;

exports[`user resolver Mutation signin should return token for correct credentials 1`] = `
Object {
  "token": "{\\"userId\\":\\"admin\\"};secret;{\\"expiresIn\\":10}",
}
`;

exports[`user resolver Query 1`] = `
Object {
  "_id": "admin",
  "credentials": Object {
    "current": Object {
      "algorithm": "dev",
      "hash": "123password",
      "salt": "123",
      "valid": true,
    },
    "store": Array [
      Object {
        "algorithm": "dev",
        "hash": "123passwordOld",
        "salt": "123",
        "valid": true,
      },
    ],
  },
}
`;
